Add "Create more" checkbox to AddTodoForm (#252)

Sachinbhanu7 opened the pull request:
Fixes #252

- Added a "Create more" checkbox to the Add Todo dialog.
- When checked, the form resets after submission instead of closing.

gabor-boros commented:
@Sachinbhanu7 first things first, thank you for your quick PR! 🎉  I just reviewed and tested the changes. It works as expected, though I would have a few notes:

* We already have fully accessible [Checkbox](https://github.com/opcotech/elemo/blob/main/web/src/components/ui/checkbox.tsx) and [Label](https://github.com/opcotech/elemo/blob/main/web/src/components/ui/label.tsx) components that we could use here
* Resetting the form is actually done (as it was before), though as values are not set for the `reset` function, it doesn't reset the value input itself.

I was not aware of the second point. It could be fixed like the following:

<details>
<summary>Code diff for reset</summary>

```diff
diff --git a/web/src/components/todo/add-todo-form.tsx b/web/src/components/todo/add-todo-form.tsx
index 8915c73..57af186 100644
--- a/web/src/components/todo/add-todo-form.tsx
+++ b/web/src/components/todo/add-todo-form.tsx
@@ -1,7 +1,7 @@
 import { zodResolver } from "@hookform/resolvers/zod";
 import { useMutation } from "@tanstack/react-query";
-import { useForm } from "react-hook-form";
 import { useState } from "react";
+import { useForm } from "react-hook-form";
 import type { z } from "zod";
 
 import { Button } from "@/components/ui/button";
@@ -48,6 +48,13 @@ interface AddTodoFormProps {
   onSuccess?: () => void;
 }
 
+const defaultValues: TodoFormValues = {
+  title: "",
+  description: undefined,
+  priority: "normal",
+  due_date: null,
+};
+
 export function AddTodoForm({
   open,
   onOpenChange,
@@ -58,10 +65,7 @@ export function AddTodoForm({
 
   const form = useForm<TodoFormValues>({
     resolver: zodResolver(todoFormSchema),
-    defaultValues: {
-      priority: "normal",
-      due_date: null,
-    },
+    defaultValues,
   });
 
   const mutation = useMutation(v1TodosCreateMutation());
@@ -81,7 +85,7 @@ export function AddTodoForm({
         onSuccess: () => {
           if (!createMore) onOpenChange(false);
           onSuccess?.();
-          form.reset();
+          form.reset(defaultValues);
           showSuccessToast(
             "Todo added successfully",
             `Todo "${values.title}" with priority "${values.priority}" added successfully`
@@ -97,7 +101,7 @@ export function AddTodoForm({
   const handleOpenChange = (newOpen: boolean) => {
     if (!newOpen) {
       // Reset form when closing
-      form.reset();
+      form.reset(defaultValues);
     }
     onOpenChange(newOpen);
   };
@@ -144,8 +148,13 @@ export function AddTodoForm({
                       placeholder="Enter todo description (optional)"
                       className="min-h-40 resize-y"
                       rows={6}
-                      {...field}
-                      value={field.value || undefined}
+                      value={field.value || ""}
+                      onChange={(e) => {
+                        const value = e.target.value;
+                        field.onChange(value === "" ? undefined : value);
+                      }}
+                      onBlur={field.onBlur}
+                      name={field.name}
                     />
                   </FormControl>
                   <FormMessage />
@@ -162,7 +171,7 @@ export function AddTodoForm({
                     <FormLabel>Priority</FormLabel>
                     <Select
                       onValueChange={field.onChange}
-                      defaultValue={field.value}
+                      value={field.value}
                     >
                       <FormControl>
                         <SelectTrigger className="w-full">
```

</details>

Would you be up for replacing the raw HTML label/input with the components and applying the diff above? If yes, then a `pnpm lint:fix && pnpm format` should fix all potential linting and formatting errors.

Sachinbhanu7 commented:
Hey @gabor-boros, thanks for the detailed review! 🙌  
- Updated the form to use `defaultValues` for consistent reset behavior  
- Replaced raw HTML `<textarea>` with controlled `Textarea` using value/onChange/onBlur/name  
- Updated `Select` to use `value` instead of `defaultValue`  

Let me know if anything else is needed!

gabor-boros reviewed (dismissed):
@Sachinbhanu7 I'm happy to help out with any guidance needed! I noticed that `pnpm format` and `pnpm lint:fix` were not executed, so I took the liberty to do that for you. Also, I applied some minor changes to align the label and adjust font style. As I cannot update your branch, could you please `git apply` the following patch? Then I can merge all your changes 😊 

<details>
<summary>patch</summary>

```patch
commit 8c2f4febff4a92e4468134f13c366fa1a3165b44
Author: Gabor Boros <gabor@opcotech.com>
Date:   Mon Jul 28 17:36:36 2025 +0200

    style: minor adjustments and fix import sort
    
    Signed-off-by: Gabor Boros <gabor@opcotech.com>

diff --git a/web/src/components/todo/add-todo-form.tsx b/web/src/components/todo/add-todo-form.tsx
index a9f14b7..2abce02 100644
--- a/web/src/components/todo/add-todo-form.tsx
+++ b/web/src/components/todo/add-todo-form.tsx
@@ -1,10 +1,11 @@
-import { useState } from "react";
-import { useForm } from "react-hook-form";
 import { zodResolver } from "@hookform/resolvers/zod";
 import { useMutation } from "@tanstack/react-query";
+import { useState } from "react";
+import { useForm } from "react-hook-form";
 import type { z } from "zod";
 
 import { Button } from "@/components/ui/button";
+import { Checkbox } from "@/components/ui/checkbox";
 import { DatePicker } from "@/components/ui/date-picker";
 import {
   Dialog,
@@ -22,6 +23,7 @@ import {
   FormMessage,
 } from "@/components/ui/form";
 import { Input } from "@/components/ui/input";
+import { Label } from "@/components/ui/label";
 import {
   Select,
   SelectContent,
@@ -31,8 +33,6 @@ import {
 } from "@/components/ui/select";
 import { Spinner } from "@/components/ui/spinner";
 import { Textarea } from "@/components/ui/textarea";
-import { Checkbox } from "@/components/ui/checkbox";
-import { Label } from "@/components/ui/label";
 import { useAuth } from "@/hooks/use-auth";
 import { v1TodosCreateMutation } from "@/lib/client/@tanstack/react-query.gen";
 import { zTodoCreate } from "@/lib/client/zod.gen";
@@ -66,9 +66,9 @@ export function AddTodoForm({
   const [createMore, setCreateMore] = useState(false);
 
   const form = useForm<TodoFormValues>({
-  resolver: zodResolver(todoFormSchema),
-  defaultValues,
-});
+    resolver: zodResolver(todoFormSchema),
+    defaultValues,
+  });
 
   const mutation = useMutation(v1TodosCreateMutation());
 
@@ -165,32 +165,29 @@ export function AddTodoForm({
             />
 
             <div className="flex gap-4">
-            <FormField
-              control={form.control}
-              name="priority"
-              render={({ field }) => (
-                <FormItem className="w-1/3">
-                  <FormLabel>Priority</FormLabel>
-                  <Select
-                    value={field.value}
-                    onValueChange={field.onChange}
-                  >
-                    <FormControl>
-                      <SelectTrigger className="w-full">
-                        <SelectValue placeholder="Select a priority" />
-                      </SelectTrigger>
-                    </FormControl>
-                    <SelectContent>
-                      <SelectItem value="normal">Normal</SelectItem>
-                      <SelectItem value="important">Important</SelectItem>
-                      <SelectItem value="urgent">Urgent</SelectItem>
-                      <SelectItem value="critical">Critical</SelectItem>
-                    </SelectContent>
-                  </Select>
-                  <FormMessage />
-                </FormItem>
-              )}
-            />
+              <FormField
+                control={form.control}
+                name="priority"
+                render={({ field }) => (
+                  <FormItem className="w-1/3">
+                    <FormLabel>Priority</FormLabel>
+                    <Select value={field.value} onValueChange={field.onChange}>
+                      <FormControl>
+                        <SelectTrigger className="w-full">
+                          <SelectValue placeholder="Select a priority" />
+                        </SelectTrigger>
+                      </FormControl>
+                      <SelectContent>
+                        <SelectItem value="normal">Normal</SelectItem>
+                        <SelectItem value="important">Important</SelectItem>
+                        <SelectItem value="urgent">Urgent</SelectItem>
+                        <SelectItem value="critical">Critical</SelectItem>
+                      </SelectContent>
+                    </Select>
+                    <FormMessage />
+                  </FormItem>
+                )}
+              />
 
               <FormField
                 control={form.control}
@@ -221,15 +218,13 @@ export function AddTodoForm({
                 <Checkbox
                   id="createMore"
                   checked={createMore}
-                  onCheckedChange={(checked) =>
-                    setCreateMore(!!checked)
-                  }
+                  onCheckedChange={(checked) => setCreateMore(!!checked)}
                 />
-                <Label htmlFor="createMore" className="text-sm">
+                <Label htmlFor="createMore" className="mb-0.5 font-normal">
                   Create more
                 </Label>
               </div>
-            
+
               <Button type="submit" disabled={mutation.isPending}>
                 {mutation.isPending ? (
                   <>
@@ -241,7 +236,6 @@ export function AddTodoForm({
                 )}
               </Button>
             </DialogFooter>
-
           </form>
         </Form>
       </DialogContent>
```

</details>

Sachinbhanu7 commented:
Hey! 😊
Thanks a ton for being so supportive and helpful throughout the process. This was actually my first-ever contribution to open source, so it really means a lot! 🙌
I’ve applied the patch as suggested — everything looks perfect now.
Looking forward to more contributions like this 🚀

gabor-boros reviewed (approved):
@Sachinbhanu7 I'm happy that you were not overwhelmed with the change requests and were enjoying the process of contributing! Also, I'm always happy to welcome anyone in the open-source world 👋 

Everything looks good for me too now; don't mind about the git commit message lint, I'm going to overwrite your commit message upon merging.

Thank you for your contribution and looking forward to more contributions from you as well! After merging the bot will mention anyway, but feel free to join Discord or Slack, so a community can be formed around Elemo.

Happy coding ahead 🚀

gabor-boros commented:
@Sachinbhanu7 No worries, [I see no linter errors](https://github.com/opcotech/elemo/actions/runs/16577642012/job/46885757398). In fact, the build was passing. The formatting of the file could have been improved by running `pnpm format`, but that's a minor thing. No need for opening a new PR for that. 😊

Sachinbhanu7 commented:
Hi @gabor-boros,

Thank you so much for the kind words and warm welcome to the open-source world!
I truly enjoyed the entire process — from understanding the codebase to implementing feedback — and learned a lot along the way.

Looking forward to contributing more to Elemo and being an active part of the community.
Just joined the Discord as well! 😊

Thanks again, and happy coding to you too! 🚀

Best regards,
Sachin Bhanushali
